Add base types for legacy crypto algorithms.

AtsushiKan opened the pull request:
cc @bartonjs

bartonjs commented on src/System.Security.Cryptography.Algorithms/tests/RC2Tests.cs:
```diff
@@ -0,0 +1,72 @@
+                    rc2.KeySize = keySize;
+                    Assert.Equal(keySize, rc2.KeySize);
+
+
```
Nit: too many blank lines

bartonjs commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace System.Security.Cryptography
+{
+    public abstract class DSA : AsymmetricAlgorithm
```
We'd probably want to add the following public members:
```C#
byte[] SignData(Stream, HashAlgorithmName);
byte[] SignData(byte[] buffer, HashAlgorithmName);
byte[] SignData(byte[] buffer, int offset, int count, HashAlgorithmName);
bool VerifyData(byte[] data, byte[] signature, HashAlgorithmName);
bool VerifyData(Stream, byte[] signature, HashAlgorithmName);
```
and the protected drivers thereof:
```C#
protected abstract byte[] HashData(byte[] data, int offset, int count, HashAlgorithmName hashAlgorithm);
protected abstract byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm);
```

But we may already be in a bind for adding those, since Desktop doesn't have them yet.

@nguerrera: Thoughts?

bartonjs commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RC2.cs:
```diff
@@ -0,0 +1,100 @@
+
+namespace System.Security.Cryptography
+{
+    public abstract class RC2 : SymmetricAlgorithm
```
RC2CryptoServiceProvider has `public bool UseSalt { get; set; }`.  Do we want/need that on the base class?  (I'm not familiar enough with RC2 to know.)

bartonjs commented:
We should get Nick's opinion on the DSA base class enhancements and contract version/management before merging; just to know what the plan forward is.

stephentoub commented on src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/Helpers.cs:
```diff
@@ -14,6 +14,11 @@ public static byte[] CloneByteArray(this byte[] src)
+        {
+            return (KeySizes[])(src.Clone());
+        }
+
```
Should we just have a helper:
```C#
public static T[] CloneArray<T>(this T[] src)
{
    return (T[])src.Clone();
}
```
rather than one method per T?

AtsushiKan replied:
A valid point but also brings in more ancillary code changes (and inconsistencies with other crypto assemblies) that I don’t want to distract from this change.

stephentoub commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DES.cs:
```diff
@@ -0,0 +1,139 @@
+
+        public static DES Create()
+        {
+            throw new NotImplementedException(SR.WorkInProgress);
```
These NIEs are going to be handled in a subsequent PR?

AtsushiKan replied:
Yeah, we have to figure out the right type-forwarding strategy for these Create() methods. The priority here is to get these base classes into the Algorithms assembly and packages
so we can add the Cng subclasses w/out friction once we switch away from P2P references.

stephentoub commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/TripleDES.cs:
```diff
@@ -0,0 +1,108 @@
+                if (value == null)
+                    throw new ArgumentNullException("value");
+
+                if (!(value.Length*8).IsLegalSize(s_legalKeySizes))
```
I personally don't like us having domain-specific extension methods on integers like this, but I realize you're not introducing this as part of your change, so ok.

stephentoub commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DES.cs:
```diff
@@ -0,0 +1,139 @@
+        public static bool IsWeakKey(byte[] rgbKey)
+        {
+            if (rgbKey == null)
+                throw new CryptographicException(SR.Cryptography_InvalidKeySize);  // Desktop compat: Strange exception for a null value, but this is what we threw in classic CLR. 
```
These types are being added purely for legacy compat, right?  If so, makes sense to keep the behavior as consistent as possible.  If, however, we believe there's non-legacy value in these types, then we should consider breaking compat here in the name of throwing the right exception.

AtsushiKan replied:
Legacy compat is indeed the only reason we want these ancient algorithms around ☺

stephentoub commented on src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RC2.cs:
```diff
@@ -0,0 +1,100 @@
+                }
+                else if (value < 40)
+                {
+                    throw new CryptographicException(SR.Cryptography_RC2_EKS40);
```
This case won't be caught by the subsequent IsLegalSize check?

AtsushiKan replied:
It does seem redundant but the original desktop code did set aside this specific error string for the  “< 40” case so I figured it was safer not to tamper.

stephentoub commented on src/System.Security.Cryptography.Algorithms/tests/DesTests.cs:
```diff
@@ -0,0 +1,147 @@
+        [Fact]
+        public static void DesIsWeakPositive()
+        {
+            foreach (byte[] key in _knownWeakKeys)
```
Consider doing this with a [Theory] and [MemberData] instead.  Then if the asserts fail, the test output will include the failed inputs.

stephentoub commented:
LGTM

AtsushiKan commented:
Removed the extranous whitespace.

Removed DSA.cs from PR - will handle this separately since we have some contract issues to haggle out (and depending on the answer, we'll probably want to refactor existing code to avoid code duplication.)

bartonjs commented:
It feels to me like we'd hold `DSAParameters` until adding DSA.  Thoughts?

AtsushiKan commented:
Well the question isn’t whether we’ll add DSA, just what methods will be on it. So I’d rather not create unnecessary new debt.

bartonjs commented:
That's fair.  It's not like we'd really debate anything on that type, and we can just leave it out of the contract until DSA makes it in (even if that's post RTM)
